From 712410471566cae25721f367d513231839070f3c Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Wed, 16 Jul 2025 19:00:49 +0200
Subject: [PATCH 901/954] feat(stm32mp2): rework SCMI driver

Factorize SCMI treatment (scmi_channel_prepare) and the error management
with timeout (scmi_rsp_wait), with protocol error (channel busy or
unexpected SCMI response size in scmi_rsp_check) or for unexpected version
in smci_proto_version_check.

The mailbox is always checked  before to sent a new message thus the
mailbox needs to be initialized (set FREE) by calling the new function
scmi_init().

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 50fd1412a011d61770ad944a9098ee1b6564cd82 feat(stm32mp2): rework SCMI driver]
Change-Id: I7092ad610f409939b558d96af7a7943c5a8b0497
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/475537
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/services/ca35ss_clk_svc.c |  31 ++--
 plat/st/stm32mp2/services/scmi_common.c    | 203 +++++++++++++++++----
 plat/st/stm32mp2/services/scmi_private.h   |   6 +-
 3 files changed, 182 insertions(+), 58 deletions(-)

diff --git a/plat/st/stm32mp2/services/ca35ss_clk_svc.c b/plat/st/stm32mp2/services/ca35ss_clk_svc.c
index 0718f51cf0..e07fc235c0 100644
--- a/plat/st/stm32mp2/services/ca35ss_clk_svc.c
+++ b/plat/st/stm32mp2/services/ca35ss_clk_svc.c
@@ -31,8 +31,6 @@
 
 #define PART_NUMBER_SUPPORT_OVERDRIVE	BIT_32(31) /* stm32mp2xxD or stm32mp2xxF */
 
-#define TIMEOUT_10MS_IN_US		10000U
-
 #if STM32MP23 || STM32MP25
 #define VOLTD_SCMI_STPMIC2_BUCK1	10U
 #define VOLTD_SCMI_STPMIC2_BUCK		VOLTD_SCMI_STPMIC2_BUCK1
@@ -298,7 +296,10 @@ static uint32_t vddcpu_get_uvolt(void)
 		return STM32_SMC_FAILED;
 	}
 
-	scmi_voltd_level_get_snd(VOLTD_SCMI_STPMIC2_BUCK);
+	if (scmi_voltd_level_get_snd(VOLTD_SCMI_STPMIC2_BUCK)) {
+		return STM32_SMC_FAILED;
+	}
+
 	return STM32_SMC_ON_GOING;
 }
 
@@ -353,8 +354,10 @@ static uint32_t vddcpu_set_uvolt(int32_t opp_idx)
 		return STM32_SMC_FAILED;
 	}
 
-	scmi_voltd_level_set_snd(VOLTD_SCMI_STPMIC2_BUCK, 0,
-				 opp.uvolt[opp_idx]);
+	if (scmi_voltd_level_set_snd(VOLTD_SCMI_STPMIC2_BUCK, 0, opp.uvolt[opp_idx])) {
+		return STM32_SMC_FAILED;
+	}
+
 	return STM32_SMC_ON_GOING;
 }
 
@@ -389,7 +392,6 @@ uint32_t ca35ss_clk_svc_setup(void)
 	int32_t opp_idx;
 	int32_t err = 0;
 	uint32_t voltd_version;
-	uint64_t timeout_ref;
 	uint32_t opp_hw_filter = get_opp_hw_filter();
 
 	/* Read the DT to find available configurations */
@@ -436,19 +438,12 @@ uint32_t ca35ss_clk_svc_setup(void)
 		}
 	}
 
-	/* Verify SCMI availlability */
-	timeout_ref = timeout_init_us(TIMEOUT_10MS_IN_US);
-	while (scmi_channel_busy())
-	{
-		if (scmi_channel_error()
-		    || timeout_elapsed(timeout_ref)) {
-			/* Could not get a free channel */
-			scmi_channel_clear();
-			ERROR("%s: Couldn't get SCMI free channel\n", __func__);
-			return STM32_SMC_FAILED;
-		}
-		udelay(10);
+	if (scmi_init() != 0) {
+		ERROR("%s: couldn't get SCMI voltd protocol\n", __func__);
+		return STM32_SMC_FAILED;
 	}
+
+	/* Verify SCMI voltage domain availability */
 	if (scmi_voltd_protocol_version(&voltd_version) != SCMI_SUCCESS) {
 		ERROR("%s: couldn't get SCMI voltd protocol\n", __func__);
 		return STM32_SMC_FAILED;
diff --git a/plat/st/stm32mp2/services/scmi_common.c b/plat/st/stm32mp2/services/scmi_common.c
index 11f7029cfb..06e47a3b1b 100644
--- a/plat/st/stm32mp2/services/scmi_common.c
+++ b/plat/st/stm32mp2/services/scmi_common.c
@@ -4,6 +4,7 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
+#include <common/debug.h>
 #include <drivers/delay_timer.h>
 #include <drivers/scmi-msg.h>
 #include <drivers/scmi.h>
@@ -41,12 +42,21 @@
 #define SCMI_MSG_VOLTAGE_LEVEL_SET	0x7U
 #define SCMI_MSG_VOLTAGE_LEVEL_GET	0x8U
 
+/* Helper macro on version of a SCMI protocol */
+#define SCMI_GET_VER_MAJOR(ver)		(((ver) >> 16) & GENMASK_32(15, 0))
+#define SCMI_GET_VER_MINOR(ver)		((ver) & GENMASK_32(15, 0))
 
-/* Helper macro to create an SCMI message header given protocol, message id and token.  */
-#define SCMI_MSG_CREATE(_protocol, _msg_id, _token)					\
-	((((_protocol) << SCMI_HEADER_PROTOCOL_SHIFT)  & SCMI_HEADER_PROTOCOL_MASK) |	\
-	 (((_msg_id) << SCMI_HEADER_MSG_ID_SHIFT) & SCMI_HEADER_MSG_ID_MASK) |		\
-	 (((_token) << SCMI_HEADER_TOKEN_SHIFT) & SCMI_HEADER_TOKEN_MASK))
+#define SCMI_VERSION(maj, min)								\
+	((((maj) & GENMASK_32(15, 0)) << 16) | ((min) & GENMASK_32(15, 0)))
+
+/* Expected SCMI protocol version for this driver */
+#define SCMI_PROTO_VER_VOLTAGE		SCMI_VERSION(1, 0)
+
+/* Check that the driver's version is same or higher than the reported SCMI version. */
+#define is_scmi_version_compatible(drv, scmi)						\
+	((SCMI_GET_VER_MAJOR(drv) > SCMI_GET_VER_MAJOR(scmi)) ||			\
+	 ((SCMI_GET_VER_MAJOR(drv) == SCMI_GET_VER_MAJOR(scmi)) &&			\
+	  (SCMI_GET_VER_MINOR(drv) <= SCMI_GET_VER_MINOR(scmi))))
 
 /* SMT Channel */
 #define SCMI_CHANNEL_FLAGS_POLL		0
@@ -109,25 +119,79 @@ bool scmi_channel_error(void)
 }
 
 /**
- * @brief Get the SCMI voltage domain protocol version.
+ * @brief Verify SCMI availlability
  *
- * @param The protocol version.
- * @return The status of the operation.
+ */
+static int32_t scmi_channel_check(void)
+{
+	uint64_t timeout_ref = timeout_init_us(TIMEOUT_10MS_IN_US);
+
+	while (scmi_channel_busy()) {
+		if (scmi_channel_error() || timeout_elapsed(timeout_ref)) {
+			/* Could not get a free channel */
+			scmi_channel_clear();
+			ERROR("%s: Couldn't get SCMI free channel\n", __func__);
+			return SCMI_ST_TIMEOUT;
+		}
+		udelay(10);
+	}
+
+	return SCMI_SUCCESS;
+}
+
+/**
+ * @brief Verify SCMI availlability
  *
  */
-int32_t scmi_voltd_protocol_version(uint32_t *version)
+static int32_t scmi_channel_prepare(uint32_t protocol_id, uint32_t msg_id, uint32_t size)
 {
-	uint64_t timeout_ref;
+	int32_t ret = scmi_channel_check();
+
+	if (ret != SCMI_SUCCESS) {
+		return ret;
+	}
 
 	shmem->channel_status = 0U;
 	shmem->channel_flags = SCMI_CHANNEL_FLAGS_POLL;
-	shmem->length = sizeof(uint32_t);
-	shmem->message_header = SCMI_MSG_CREATE(SCMI_PROTOCOL_ID_VOLTAGE_DOMAIN,
-						SCMI_MSG_PROTOCOL_VERSION, 0U);
+	shmem->length = (1 + size) * sizeof(uint32_t); /* heaader and payload */
+	shmem->message_header =
+		(((protocol_id) << SCMI_HEADER_PROTOCOL_SHIFT)  & SCMI_HEADER_PROTOCOL_MASK) |
+		(((msg_id) << SCMI_HEADER_MSG_ID_SHIFT) & SCMI_HEADER_MSG_ID_MASK) |
+		(((0U) << SCMI_HEADER_TOKEN_SHIFT) & SCMI_HEADER_TOKEN_MASK);
 
-	scmi_ring_doorbell();
+	return SCMI_SUCCESS;
+}
+
+/**
+ * @brief Check the SCMI answer
+ *
+ */
+static int32_t scmi_rsp_check(uint32_t size)
+{
+	scmi_channel_clear();
+
+	if (shmem->length == 2 * sizeof(uint32_t)) {
+		/* Return the first paylod value = status. For example, NOT SUPPORTED */
+		return (int32_t)shmem->payload[0];
+	} else if (shmem->length != ((1 + size) * sizeof(uint32_t))) {
+		INFO("SCMI 0x%x msg %u: size %u, expected %lu\n",
+		     SCMI_MSG_GET_PROTOCOL(shmem->message_header),
+		     SCMI_MSG_GET_MSG_ID(shmem->message_header),
+		     shmem->length, (size + 1) * sizeof(uint32_t));
+		return SCMI_COMMS_ERROR;
+	}
+
+	return SCMI_SUCCESS;
+}
+
+/**
+ * @brief Wait the SCMI answer with expected size
+ *
+ */
+static int32_t scmi_rsp_wait(uint32_t timeout_us, uint32_t size)
+{
+	uint64_t timeout_ref = timeout_init_us(timeout_us);
 
-	timeout_ref = timeout_init_us(TIMEOUT_10MS_IN_US);
 	while (scmi_channel_busy() && !scmi_channel_error())
 	{
 		if (timeout_elapsed(timeout_ref)) {
@@ -141,31 +205,82 @@ int32_t scmi_voltd_protocol_version(uint32_t *version)
 		return SCMI_COMMS_ERROR;
 	}
 
-	scmi_channel_clear();
+	return scmi_rsp_check(size);
+}
 
-	if (shmem->length != (3U * sizeof(uint32_t))) {
-		return SCMI_COMMS_ERROR;
+/* Query the protocol message attributes for a SCMI protocol */
+static int32_t scmi_proto_version(uint32_t proto_id, uint32_t *version)
+{
+	int32_t ret = scmi_channel_prepare(proto_id, SCMI_MSG_PROTOCOL_VERSION, 0U);
+
+	if (ret != SCMI_SUCCESS) {
+		return ret;
 	}
 
+	scmi_ring_doorbell();
+
+	ret = scmi_rsp_wait(TIMEOUT_10MS_IN_US, 2U);
+	if (ret != SCMI_SUCCESS) {
+		return ret;
+	}
 	*version = shmem->payload[1];
+
 	return (int32_t)shmem->payload[0]; /* status */
 }
 
+static int32_t smci_proto_version_check(uint32_t proto_id, uint32_t expected)
+{
+	uint32_t version;
+	int32_t ret;
+
+	ret = scmi_proto_version(proto_id, &version);
+	if (ret == SCMI_SUCCESS) {
+		if (!is_scmi_version_compatible(version, expected)) {
+			WARN("SCMI protocol 0x%x version 0x%x, expected 0x%x\n",
+			     proto_id, version, expected);
+			return SCMI_OUT_OF_RANGE;
+		}
+	} else if (ret == SCMI_NOT_SUPPORTED) {
+		/* Not supported is a valid error */
+		ret = SCMI_SUCCESS;
+	} else {
+		ERROR("SCMI protocol 0x%x: init error %d\n", proto_id, ret);
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Get the SCMI voltage domain protocol version.
+ *
+ * @param The protocol version.
+ * @return The status of the operation.
+ *
+ */
+int32_t scmi_voltd_protocol_version(uint32_t *version)
+{
+	return scmi_proto_version(SCMI_PROTOCOL_ID_VOLTAGE_DOMAIN, version);
+}
+
 /**
  * @brief Send a request to get the voltage level of a domain.
  *
  * @param domain_id The ID of the voltage domain.
  *
  */
-void scmi_voltd_level_get_snd(uint32_t domain_id)
+int32_t scmi_voltd_level_get_snd(uint32_t domain_id)
 {
-	shmem->channel_status = 0U;
-	shmem->channel_flags = SCMI_CHANNEL_FLAGS_POLL;
-	shmem->length = 2U * sizeof(uint32_t);
-	shmem->message_header = SCMI_MSG_CREATE(SCMI_PROTOCOL_ID_VOLTAGE_DOMAIN,
-						SCMI_MSG_VOLTAGE_LEVEL_GET, 0U);
+	int32_t ret = scmi_channel_prepare(SCMI_PROTOCOL_ID_VOLTAGE_DOMAIN,
+					   SCMI_MSG_VOLTAGE_LEVEL_GET, 1U);
+	if (ret != SCMI_SUCCESS) {
+		return ret;
+	}
+
 	shmem->payload[0] = domain_id;
+
 	scmi_ring_doorbell();
+
+	return SCMI_SUCCESS;
 }
 
 /**
@@ -177,10 +292,10 @@ void scmi_voltd_level_get_snd(uint32_t domain_id)
  */
 int32_t scmi_voltd_level_get_rcv(int32_t *voltage_level)
 {
-	scmi_channel_clear();
+	int32_t ret = scmi_rsp_check(2U);
 
-	if (shmem->length != (3U * sizeof(uint32_t))) {
-		return SCMI_COMMS_ERROR;
+	if (ret != SCMI_SUCCESS) {
+		return ret;
 	}
 
 	*voltage_level = shmem->payload[1];
@@ -195,19 +310,21 @@ int32_t scmi_voltd_level_get_rcv(int32_t *voltage_level)
  * @param voltage_level The voltage level to set.
  *
  */
-void scmi_voltd_level_set_snd(uint32_t domain_id, uint32_t flags,
-			      int32_t voltage_level)
+int32_t scmi_voltd_level_set_snd(uint32_t domain_id, uint32_t flags, int32_t voltage_level)
 {
-	shmem->channel_status = 0U;
-	shmem->channel_flags = SCMI_CHANNEL_FLAGS_POLL;
-	shmem->length = 4U * sizeof(uint32_t);
-	shmem->message_header = SCMI_MSG_CREATE(SCMI_PROTOCOL_ID_VOLTAGE_DOMAIN,
-						SCMI_MSG_VOLTAGE_LEVEL_GET, 0U);
+	int32_t ret = scmi_channel_prepare(SCMI_PROTOCOL_ID_VOLTAGE_DOMAIN,
+					   SCMI_MSG_VOLTAGE_LEVEL_SET, 3U);
+	if (ret != SCMI_SUCCESS) {
+		return ret;
+	}
+
 	shmem->payload[0] = domain_id;
 	shmem->payload[1] = flags;
 	shmem->payload[2] = (uint32_t)voltage_level;
 
 	scmi_ring_doorbell();
+
+	return SCMI_SUCCESS;
 }
 
 /**
@@ -218,11 +335,23 @@ void scmi_voltd_level_set_snd(uint32_t domain_id, uint32_t flags,
  */
 int32_t scmi_voltd_level_set_rcv()
 {
-	scmi_channel_clear();
+	int32_t ret = scmi_rsp_check(1U);
 
-	if (shmem->length != (2U * sizeof(uint32_t))) {
-		return SCMI_COMMS_ERROR;
+	if (ret != SCMI_SUCCESS) {
+		return ret;
 	}
 
 	return (int32_t)shmem->payload[0]; /* status */
 }
+
+int32_t scmi_init(void)
+{
+	int32_t ret;
+
+	/* Initialize channel */
+	scmi_channel_clear();
+
+	ret = smci_proto_version_check(SCMI_PROTOCOL_ID_VOLTAGE_DOMAIN, SCMI_PROTO_VER_VOLTAGE);
+
+	return ret;
+}
diff --git a/plat/st/stm32mp2/services/scmi_private.h b/plat/st/stm32mp2/services/scmi_private.h
index 878d732bf9..d71f7d4eb3 100644
--- a/plat/st/stm32mp2/services/scmi_private.h
+++ b/plat/st/stm32mp2/services/scmi_private.h
@@ -12,15 +12,15 @@
 
 #define SCMI_ST_TIMEOUT                 (-128)
 
+int32_t scmi_init(void);
 void scmi_channel_clear(void);
 bool scmi_channel_busy(void);
 bool scmi_channel_error(void);
 
 int32_t scmi_voltd_protocol_version(uint32_t *version);
-void scmi_voltd_level_get_snd(uint32_t domain_id);
+int32_t scmi_voltd_level_get_snd(uint32_t domain_id);
 int32_t scmi_voltd_level_get_rcv(int32_t *voltage_level);
-void scmi_voltd_level_set_snd(uint32_t domain_id, uint32_t flags,
-                              int32_t voltage_level);
+int32_t scmi_voltd_level_set_snd(uint32_t domain_id, uint32_t flags, int32_t voltage_level);
 int32_t scmi_voltd_level_set_rcv();
 
 #endif /* STM32MP2_SCMI_PRIVATE */
\ No newline at end of file
-- 
2.53.0

